compiler/ilgen/StatementNames.hpp

```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_STATEMENTNAMES_INCL
+#define OMR_STATEMENTNAMES_INCL
+
+#include <cstdint>
+
+namespace OMR
+{
+
+namespace StatementName {
+
+static const int16_t VERSION_MAJOR              = 0;
```

Comment:
these values should be pulled from the jitbuilder api definition rather than being defined here, somehow. @Leonardo2718 what's the best way to do that?

Reply:
Well, I'm not sure about the "best way" :stuck_out_tongue: .

It seems to me like whole file can/should be generated from the JitBuilder API description.

A few lines of Python should do the trick; something like (not thoroughly test, ideally with less hardcoded strings and a better design):

```python
import genutils

if __name__ == "__main__":
    api_file = "jitbuilder.api.json"
    dest_file = "StatementNames.hpp"

    with open(api_file) as api_src:
        api_description = genutils.APIDescription.load_json_file(api_src)

    with open(dest_file, "w") as writer:
        writer.write("#ifndef OMR_STATEMENTNAMES_INCL\n")
        writer.write("#define OMR_STATEMENTNAMES_INCL\n")

        writer.write("#include <cstdint>\n")

        writer.write("namespace OMR {\n")
        writer.write("namespace StatementNames {\n")

        version = api_description.description["version"] # need to add methods for this to APIDescription
        writer.write("static const int16_t VERSION_MAJOR = {};\n".format(version["major"]))
        writer.write("static const int16_t VERSION_MINOR = {};\n".format(version["minor"]))
        writer.write("static const int16_t VERSION_PATCH = {};\n".format(version["patch"]))

        writer.write('static const char * const RECORDER_SIGNATURE = "JBIL";\n')
        writer.write('static const char * const JBIL_COMPLETE      = "Done";\n')

        for service in api_description.get_class_by_name("MethodBuilder").services():
            name = service.name()
            writer.write('static const char * const STATEMENT_{} = "{}";\n'.format(name.upper(), name))
        for service in api_description.get_class_by_name("IlBuilder").services():
            name = service.name()
            writer.write('static const char * const STATEMENT_{} = "{}";\n'.format(name.upper(), name))

        writer.write('static const char * const STATEMENT_ID16BIT = "ID16BIT";\n')
        writer.write('static const char * const STATEMENT_ID32BIT = "ID32BIT";\n')

        writer.write("} // namespace StatementNames\n")
        writer.write("} // namespace OMR\n")

        writer.write("#endif // OMR_STATEMENTNAMES_INCL\n")
```

Then, it's just a matter of adding a few lines to the build files to run this code. :slightly_smiling_face:

Reply:
yeah, that's a good point about the services being autogenerated. I do like that better. Thanks for the suggestion!

Reply:
Maybe we leave the improvement of this part of the infrastructure to a second PR, though. For now, @bragaigor, let's just leave the major.minor.revision levels hard-coded and we'll fix it in a subsequent PR.